

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>General Usage &mdash; xpresstools 0.0.1-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Advanced Usage" href="advanced-usage.html" />
    <link rel="prev" title="Quickstart" href="quickstart.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> xpresstools
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">General Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#important-file-naming-conventions">Important File Naming Conventions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#riboseq-module"><code class="docutils literal notranslate"><span class="pre">riboseq</span></code> module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#additional-information">Additional information:</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#rnaseq-module"><code class="docutils literal notranslate"><span class="pre">rnaseq</span></code> module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Additional information:</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#trim-module"><code class="docutils literal notranslate"><span class="pre">trim</span></code> module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#align-module"><code class="docutils literal notranslate"><span class="pre">align</span></code> module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#quality-module"><code class="docutils literal notranslate"><span class="pre">quality</span></code> module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id2">Additional information:</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#rrna-prober-module"><code class="docutils literal notranslate"><span class="pre">rrna_prober</span></code> module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#gene-dictionary-module"><code class="docutils literal notranslate"><span class="pre">gene_dictionary</span></code> module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#diffex-module"><code class="docutils literal notranslate"><span class="pre">diffex</span></code> module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#truncate-module"><code class="docutils literal notranslate"><span class="pre">truncate</span></code> module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#data-output">Data Output</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="advanced-usage.html">Advanced Usage</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">xpresstools</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>General Usage</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/general-usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="general-usage">
<h1>General Usage<a class="headerlink" href="#general-usage" title="Permalink to this headline">¶</a></h1>
<p>The purpose of RiboPipe is to automate the alignment, quality control, and initial analysis of ribosome profiling and other short, single-end read data. It is intended that
input data is a directory of .fastq formatted files. However, when using the intermediate submodules, such as <code class="xref py py-data docutils literal notranslate"><span class="pre">align</span></code> or <code class="xref py py-data docutils literal notranslate"><span class="pre">quality</span></code>, input will vary and is explicated
in the <code class="xref py py-data docutils literal notranslate"><span class="pre">--help</span></code> menu for each submodule.</p>
<p>RiboPipe was created with the novice bioinformatician in mind, and the documentation written assuming no background in using Command Line or programming. As such, instructions
have been created with as much detail as possible. Additionally, <a href="#id13"><span class="problematic" id="id14">`walkthrough videos &lt;&gt;`_</span></a> are available. In cases where RiboPipe is being used on a cloud computing cluster, or
Linux machine with Singularity installed, a singularity container image can be used to avoid the installation process. Additionally, alignment references have been created for a
variety of model organisms using current builds to help users avoid this step. Details on how to incorporate a newly created reference not already provided within the RiboPipe
infrastructure can be found <a class="reference external" href="advanced-usage.html">here</a>.</p>
<p>RiboPipe can be run essentially from beginning to end as a pipeline, or as individual submodules. We will describe each option in detail below.</p>
<div class="section" id="important-file-naming-conventions">
<h2>Important File Naming Conventions<a class="headerlink" href="#important-file-naming-conventions" title="Permalink to this headline">¶</a></h2>
<p>In order for many of the RiboPipe functions to perform properly and for the output to be reliable after alignment (except for generation of a raw counts table), file naming conventions must be followed.</p>
<ol class="arabic simple">
<li>Download your raw sequence data and place in a folder – this folder should contain all the sequence data and nothing else.</li>
<li>Make sure files follow a pattern naming scheme. For example, if you had 3 genetic backgrounds of ribosome profiling data, the naming scheme would go as follows:</li>
</ol>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ExperimentName_BackgroundA_FP.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_BackgroundA_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_BackgroundB_FP.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_BackgroundB_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_BackgroundC_FP.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_BackgroundC_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li>If the sample names are replicates, their sample number needs to be indicated.</li>
<li>If you want the final count table to be in a particular order and the samples ordered that way are not alphabetically, append a letter in front of the sample name to force this ordering.</li>
</ol>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ExperimentName_a_WT_FP.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_a_WT_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_b_exType_FP.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_b_exType_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
</pre></div>
</div>
<ol class="arabic simple" start="5">
<li>If you have replicates:</li>
</ol>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ExperimentName_a_WT_1_FP.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_a_WT_1_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_a_WT_2_FP.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_a_WT_2_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_b_exType_1_FP.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_b_exType_1_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_b_exType_2_FP.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_b_exType_2_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
</pre></div>
</div>
<ol class="arabic simple" start="6">
<li>If you are just running RNAseq files through the pipeline, you only need the RNA samples in your input directory and specify the rnaseq module:</li>
</ol>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ExperimentName_a_WT_1_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_a_WT_2_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_b_exType_1_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>
ExperimentName_b_exType_2_RNA.fastq<span class="o">(</span>.qz<span class="o">)</span>

ribopipe rnaseq -i input_directory ...
</pre></div>
</div>
</div>
<div class="section" id="riboseq-module">
<h2><code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> module<a class="headerlink" href="#riboseq-module" title="Permalink to this headline">¶</a></h2>
<p>Pipeline for handling raw ribosome profiling sequence data. Runs quality and adaptor trimming, alignment, quality control, and formatting on a directory of raw ribosome profiling
sequence data.</p>
<p>The <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> module can be run as follows:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ ribopipe riboseq -i <span class="nv">$DATA_PATH</span>/raw_ingolia/ -o <span class="nv">$DATA_PATH</span>/out_ingolia/ -r yeast -e ingolia_2015 <span class="se">\</span>
                   -s a_WT_DED1_1_15deg b_WT_DED1_2_15deg c_ded1_cs_1_15deg d_ded1_cs_2_15deg <span class="se">\</span>
                   e_WT_DED1_1_37deg f_WT_DED1_2_37deg g_ded1_ts_1_37deg h_ded_ts_2_37deg <span class="se">\</span>
                   i_WT_TIF1_1_30deg j_WT_TIF1_2_30deg k_tif1_ts_1_30deg l_tif1_ts_2_30deg <span class="se">\</span>
                   m_WT_TIF1_1_37deg n_WT_TIF1_2_37deg o_tif1_ts_1_37deg p_tif1_ts_2_37deg <span class="se">\</span>
                   -p HISAT2 -a CTGTAGGCACCATCAAT --platform ILLUMINA --count_cutoff <span class="m">32</span>
</pre></div>
</div>
<p>An explanation of the <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> submodule arguments can be found in the following tables:</p>
<table border="1" class="colwidths-given docutils" id="id3">
<caption><span class="caption-text"><code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> required arguments</span><a class="headerlink" href="#id3" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="41%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Argument</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">INPUT,</span> <span class="pre">--input</span> <span class="pre">INPUT</span></code></td>
<td>Specify full PATH to input directory</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">OUTPUT,</span> <span class="pre">--output</span> <span class="pre">OUTPUT</span></code></td>
<td>Specify full PATH to output directory</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-r</span> <span class="pre">&lt;yeast&gt;,</span> <span class="pre">&lt;human&gt;,</span> <span class="pre">&lt;mouse&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--reference</span> <span class="pre">&lt;yeast&gt;,</span> <span class="pre">&lt;human&gt;,</span> <span class="pre">&lt;mouse&gt;</span></code></td>
<td>Specify model organism used for experiments. Pipeline will align a current, curated reference</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-e</span> <span class="pre">&lt;string&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--experiment</span> <span class="pre">&lt;string&gt;</span></code></td>
<td>Provide experiment name to prepend to output files</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-s</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--samples</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code></td>
<td>Space delimited list of samples in order. If replicates are included, indicate number in sample name to delineate.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-a</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--adaptor</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code></td>
<td>Sequence of 3’ linker (only supports one 3’ linker currently) (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">AACTGTAGGCACCATCAAT</span></code>). If no adaptor was used, specify <code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code>’</td>
</tr>
</tbody>
</table>
<table border="1" class="colwidths-given docutils" id="id4">
<caption><span class="caption-text"><code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> optional arguments</span><a class="headerlink" href="#id4" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="41%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Argument</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span> <span class="pre">&lt;int&gt;,</span> <span class="pre">--max_processors</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Number of max processors pipeline can use for multiprocessing tasks (default: No limit)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-f,</span> <span class="pre">--footprints_only</span></code></td>
<td>Select this option if ONLY providing raw footprint sequence data</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--min_overlap</span> <span class="pre">&lt;integer&gt;</span></code></td>
<td>Minimum number of bases that must match on a side to combine sequences for <code class="xref py py-data docutils literal notranslate"><span class="pre">rrna_prober</span></code></td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-p</span> <span class="pre">&lt;HISAT2&gt;,</span> <span class="pre">&lt;STAR&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--program</span> <span class="pre">&lt;HISAT2&gt;,</span> <span class="pre">&lt;STAR&gt;</span></code></td>
<td>Alignment software to be used to align reads to reference (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">STAR</span></code>)</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--read_length_min</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Minimum read length threshold to keep for reads (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">11</span></code>)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--read_length_max</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Maximum read length threshold to keep for reads (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">50</span></code>)</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--read_quality</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>PHRED read quality threshold (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">28</span></code>)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--platform</span> <span class="pre">&lt;SANGER&gt;,</span> <span class="pre">&lt;ILLUMINA&gt;</span></code></td>
<td>Sequencing platform used (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">ILLUMINA</span></code>)</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--full_genome</span></code></td>
<td>Add this option to map reads to full genome. If not given, will not count any read mapping to the first 45 nt of transcripts for ribosome profiling</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--count_cutoff</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Minimum counts threshold. Will remove any row in the final count tables if any sample does not meet this cutoff threshold</td>
</tr>
</tbody>
</table>
<div class="section" id="additional-information">
<h3>Additional information:<a class="headerlink" href="#additional-information" title="Permalink to this headline">¶</a></h3>
<p><code class="xref py py-data docutils literal notranslate"><span class="pre">-s</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--samples</span></code>: For ribosome profiling, do not differentiate between footprint and RNA samples.</p>
<p><code class="xref py py-data docutils literal notranslate"><span class="pre">--full_genome</span></code>: It is recommended to NOT include this option as ribosome profiling data for this region is often unreliable.</p>
<p>Follow up with the <code class="xref py py-data docutils literal notranslate"><span class="pre">diffex</span></code> submodule for differential expression analysis.</p>
</div>
</div>
<div class="section" id="rnaseq-module">
<h2><code class="xref py py-data docutils literal notranslate"><span class="pre">rnaseq</span></code> module<a class="headerlink" href="#rnaseq-module" title="Permalink to this headline">¶</a></h2>
<p>Similar to the <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> module, but tuned for small RNAseq (i.e. anything single-end under 100 bps). It is expected that in the future, RiboPipe will be able to perform
automated paired-end and genome sequencing assembly and alignment.</p>
<p>The <code class="xref py py-data docutils literal notranslate"><span class="pre">rnaseq</span></code> module can be run as follows:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ ribopipe rnaseq -i <span class="nv">$DATA_PATH</span>/raw_dang_berger/ -o <span class="nv">$DATA_PATH</span>/out_dang_berger/ -r yeast -e dang_berger_2014 <span class="se">\</span>
                  -s WT_NR_A WT_NR_B WT_CR_A WT_CR_B <span class="se">\</span>
                  -p HISAT2 -a TGGAATTCTCGGGTGCCAAGG --platform ILLUMINA --replicates
</pre></div>
</div>
<p>An explanation of the <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> submodule commands can be found in the following tables:</p>
<table border="1" class="colwidths-given docutils" id="id5">
<caption><span class="caption-text"><code class="xref py py-data docutils literal notranslate"><span class="pre">rnaseq</span></code> required arguments</span><a class="headerlink" href="#id5" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="41%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Argument</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">INPUT,</span> <span class="pre">--input</span> <span class="pre">INPUT</span></code></td>
<td>Specify full PATH to input directory</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">OUTPUT,</span> <span class="pre">--output</span> <span class="pre">OUTPUT</span></code></td>
<td>Specify full PATH to output directory</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-r</span> <span class="pre">&lt;yeast&gt;,</span> <span class="pre">&lt;human&gt;,</span> <span class="pre">&lt;mouse&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--reference</span> <span class="pre">&lt;yeast&gt;,</span> <span class="pre">&lt;human&gt;,</span> <span class="pre">&lt;mouse&gt;</span></code></td>
<td>Specify model organism used for experiments. Pipeline will align a current, curated reference</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-e</span> <span class="pre">&lt;string&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--experiment</span> <span class="pre">&lt;string&gt;</span></code></td>
<td>Provide experiment name to prepend to output files</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-s</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--samples</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code></td>
<td>Space delimited list of samples in order. If replicates are included, indicate number in sample name to delineate.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-a</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--adaptor</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code></td>
<td>Sequence of 3’ linker (only supports one 3’ linker currently) (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">AACTGTAGGCACCATCAAT</span></code>). If no adaptor was used, specify <code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code>’</td>
</tr>
</tbody>
</table>
<table border="1" class="colwidths-given docutils" id="id6">
<caption><span class="caption-text"><code class="xref py py-data docutils literal notranslate"><span class="pre">rnaseq</span></code> optional arguments</span><a class="headerlink" href="#id6" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="41%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Argument</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span> <span class="pre">&lt;int&gt;,</span> <span class="pre">--max_processors</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Number of max processors pipeline can use for multiprocessing tasks (default: No limit)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--replicates</span></code></td>
<td>Select this option if samples are replicates (do not use if 3+ replicates).</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-p</span> <span class="pre">&lt;HISAT2&gt;,</span> <span class="pre">&lt;STAR&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--program</span> <span class="pre">&lt;HISAT2&gt;,</span> <span class="pre">&lt;STAR&gt;</span></code></td>
<td>Alignment software to be used to align reads to reference (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">STAR</span></code>)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--read_length_min</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Minimum read length threshold to keep for reads (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">11</span></code>)</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--read_length_max</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Maximum read length threshold to keep for reads (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">50</span></code>)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--read_quality</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>PHRED read quality threshold (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">28</span></code>)</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--platform</span> <span class="pre">&lt;SANGER&gt;,</span> <span class="pre">&lt;ILLUMINA&gt;</span></code></td>
<td>Sequencing platform used (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">ILLUMINA</span></code>)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--full_genome</span></code></td>
<td>Add this option to map reads to full genome. If not given, will only map reads to transcripts.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--count_cutoff</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Minimum counts threshold. Will remove any row in the final count tables if any sample does not meet this cutoff threshold</td>
</tr>
</tbody>
</table>
<div class="section" id="id1">
<h3>Additional information:<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p><code class="xref py py-data docutils literal notranslate"><span class="pre">--replicates</span></code>: Make sure when passing the argument to list samples these are sorted so replicates are next to each other in this list</p>
<p>Follow up with the <code class="xref py py-data docutils literal notranslate"><span class="pre">diffex</span></code> submodule for differential expression analysis.</p>
</div>
</div>
<div class="section" id="trim-module">
<h2><code class="xref py py-data docutils literal notranslate"><span class="pre">trim</span></code> module<a class="headerlink" href="#trim-module" title="Permalink to this headline">¶</a></h2>
<p>Perform only trimming and quality control steps, outputting general read information. Will output trimmed reads ready for alignment.</p>
<p>The <code class="xref py py-data docutils literal notranslate"><span class="pre">trim</span></code> module can be run as follows:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ ribopipe trim -i <span class="nv">$DATA_PATH</span>/input_files/ -o <span class="nv">$DATA_PATH</span>/output_files/ <span class="se">\</span>
                  -a TGGAATTCTCGGGTGCCAAGG --platform ILLUMINA
</pre></div>
</div>
<p>An explanation of the <code class="xref py py-data docutils literal notranslate"><span class="pre">trim</span></code> submodule commands can be found in the following tables:</p>
<table border="1" class="colwidths-given docutils" id="id7">
<caption><span class="caption-text"><code class="xref py py-data docutils literal notranslate"><span class="pre">trim</span></code> required arguments</span><a class="headerlink" href="#id7" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="41%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Argument</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">INPUT,</span> <span class="pre">--input</span> <span class="pre">INPUT</span></code></td>
<td>Specify full PATH to input directory</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">OUTPUT,</span> <span class="pre">--output</span> <span class="pre">OUTPUT</span></code></td>
<td>Specify full PATH to output directory</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-a</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--adaptor</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code></td>
<td>Sequence of 3’ linker (only supports one 3’ linker currently) (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">AACTGTAGGCACCATCAAT</span></code>). If no adaptor was used, specify <code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code>’</td>
</tr>
</tbody>
</table>
<table border="1" class="colwidths-given docutils" id="id8">
<caption><span class="caption-text"><code class="xref py py-data docutils literal notranslate"><span class="pre">trim</span></code> optional arguments</span><a class="headerlink" href="#id8" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="41%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Argument</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span> <span class="pre">&lt;int&gt;,</span> <span class="pre">--max_processors</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Number of max processors pipeline can use for multiprocessing tasks (default: No limit)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--read_length_min</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Minimum read length threshold to keep for reads (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">11</span></code>)</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--read_length_max</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Maximum read length threshold to keep for reads (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">50</span></code>)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--read_quality</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>PHRED read quality threshold (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">28</span></code>)</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--platform</span> <span class="pre">&lt;SANGER&gt;,</span> <span class="pre">&lt;ILLUMINA&gt;</span></code></td>
<td>Sequencing platform used (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">ILLUMINA</span></code>)</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="align-module">
<h2><code class="xref py py-data docutils literal notranslate"><span class="pre">align</span></code> module<a class="headerlink" href="#align-module" title="Permalink to this headline">¶</a></h2>
<p>Perform only alignment of input files – assumes files have already been trimmed if required. Will output alignment files and genome browser compatible files.</p>
<p>The <code class="xref py py-data docutils literal notranslate"><span class="pre">align</span></code> module can be run as follows:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ ribopipe align -i <span class="nv">$DATA_PATH</span>/input_files/ -o <span class="nv">$DATA_PATH</span>/output_files/ -r yeast -e align_only_test <span class="se">\</span>
                  -s sample1_FP sample1_RNA sample2_FP sample2_RNA
                  -a TGGAATTCTCGGGTGCCAAGG --platform ILLUMINA
</pre></div>
</div>
<p>An explanation of the <code class="xref py py-data docutils literal notranslate"><span class="pre">align</span></code> submodule commands can be found in the following tables:</p>
<table border="1" class="colwidths-given docutils" id="id9">
<caption><span class="caption-text"><code class="xref py py-data docutils literal notranslate"><span class="pre">align</span></code> required arguments</span><a class="headerlink" href="#id9" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="41%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Argument</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">INPUT,</span> <span class="pre">--input</span> <span class="pre">INPUT</span></code></td>
<td>Specify full PATH to input directory</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">OUTPUT,</span> <span class="pre">--output</span> <span class="pre">OUTPUT</span></code></td>
<td>Specify full PATH to output directory</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-r</span> <span class="pre">&lt;yeast&gt;,</span> <span class="pre">&lt;human&gt;,</span> <span class="pre">&lt;mouse&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--reference</span> <span class="pre">&lt;yeast&gt;,</span> <span class="pre">&lt;human&gt;,</span> <span class="pre">&lt;mouse&gt;</span></code></td>
<td>Specify model organism used for experiments. Pipeline will align a current, curated reference</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-e</span> <span class="pre">&lt;string&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--experiment</span> <span class="pre">&lt;string&gt;</span></code></td>
<td>Provide experiment name to prepend to output files</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-s</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--samples</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code></td>
<td>Space delimited list of samples in order. If replicates are included, indicate number in sample name to delineate.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">t</span> <span class="pre">&lt;riboseq&gt;,</span> <span class="pre">&lt;rnaseq&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--type</span> <span class="pre">&lt;riboseq&gt;,</span> <span class="pre">&lt;rnaseq&gt;</span></code></td>
<td>Sequencing type – <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> for ribosome profiling or <code class="xref py py-data docutils literal notranslate"><span class="pre">rnaseq</span></code> for single-end short read sequence data.</td>
</tr>
</tbody>
</table>
<table border="1" class="colwidths-given docutils" id="id10">
<caption><span class="caption-text"><code class="xref py py-data docutils literal notranslate"><span class="pre">align</span></code> optional arguments</span><a class="headerlink" href="#id10" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="41%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Argument</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span> <span class="pre">&lt;int&gt;,</span> <span class="pre">--max_processors</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Number of max processors pipeline can use for multiprocessing tasks (default: No limit)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-p</span> <span class="pre">&lt;HISAT2&gt;,</span> <span class="pre">&lt;STAR&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--program</span> <span class="pre">&lt;HISAT2&gt;,</span> <span class="pre">&lt;STAR&gt;</span></code></td>
<td>Alignment software to be used to align reads to reference (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">STAR</span></code>)</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--full_genome</span></code></td>
<td>Add this argument if you wish to align to the full genome. See notes in <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> and <code class="xref py py-data docutils literal notranslate"><span class="pre">rnaseq</span></code> sections above for more information.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">--count_cutoff</span> <span class="pre">&lt;int&gt;</span></code></td>
<td>Minimum counts threshold. Will remove any row in the final count tables if any sample does not meet this cutoff threshold</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="quality-module">
<h2><code class="xref py py-data docutils literal notranslate"><span class="pre">quality</span></code> module<a class="headerlink" href="#quality-module" title="Permalink to this headline">¶</a></h2>
<p>Takes a table of raw counts generates paired scatter plots.</p>
<p>The <code class="xref py py-data docutils literal notranslate"><span class="pre">quality</span></code> module can be run as follows:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ ribopipe quality -i <span class="nv">$DATA_PATH</span>/raw_counts.csv -o <span class="nv">$DATA_PATH</span>/output_location/ <span class="se">\</span>
                  -t riboseq
</pre></div>
</div>
<p>An explanation of the <code class="xref py py-data docutils literal notranslate"><span class="pre">quality</span></code> submodule commands can be found in the following tables:</p>
<table border="1" class="colwidths-given docutils" id="id11">
<caption><span class="caption-text"><code class="xref py py-data docutils literal notranslate"><span class="pre">quality</span></code> required arguments</span><a class="headerlink" href="#id11" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="41%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Argument</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">INPUT,</span> <span class="pre">--input</span> <span class="pre">INPUT</span></code></td>
<td>Input table (must be .csv file) of raw counts</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">OUTPUT,</span> <span class="pre">--output</span> <span class="pre">OUTPUT</span></code></td>
<td>Specify full PATH to output directory</td>
</tr>
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-t</span> <span class="pre">&lt;riboseq&gt;,</span> <span class="pre">&lt;rnaseq&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--type</span> <span class="pre">&lt;riboseq&gt;,</span> <span class="pre">&lt;rnaseq&gt;</span></code></td>
<td>Sequencing type – <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> for ribosome profiling or <code class="xref py py-data docutils literal notranslate"><span class="pre">rnaseq</span></code> for single-end short read sequence data.</td>
</tr>
</tbody>
</table>
<div class="section" id="id2">
<h3>Additional information:<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>Table must have samples set as columns and must be ordered as sample1_FP, sample1_RNA, sample2_FP, sample2_RNA, etc. or as sample1_rep1_RNA, sample1_rep2_RNA, sample2_rep1_RNA, sample2_rep2_RNA.</p>
</div>
</div>
<div class="section" id="rrna-prober-module">
<h2><code class="xref py py-data docutils literal notranslate"><span class="pre">rrna_prober</span></code> module<a class="headerlink" href="#rrna-prober-module" title="Permalink to this headline">¶</a></h2>
<p>As the bulk of total cellular RNA consists mostly of ribosomal RNA that is often not of interest and crowds out measurement of other RNAs, an rRNA depletion step is commonly performed before
creating a cDNA sequencing library. Several commercial kits are available that can effectively deplete samples of full length rRNAs. However, inherent in ribosome profiling protocols, samples are
treated with an RNase, which degrades full length rRNAs, making them more difficult to deplete before creating the cDNA library. One way to better deplete these rRNA fragments in ribosome profiling
samples is to create biotinylated RNA probes for overrepresented rRNA fragments to extract these fragments (see this <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/28579404">paper</a> for a recent discussion).
<code class="xref py py-data docutils literal notranslate"><span class="pre">rrna_prober</span></code> automates the identification of overrepresented species and outputs a table with these sequences ordered by abundance for ease of probe generation. This table is output in stdout
in this module, or as a <code class="xref py py-data docutils literal notranslate"><span class="pre">txt</span></code> file in the <code class="xref py py-data docutils literal notranslate"><span class="pre">highlights/</span></code> output folder when running the full <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> submodule.</p>
<p>The <code class="xref py py-data docutils literal notranslate"><span class="pre">rrna_prober</span></code> module can be run as follows:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ ribopipe rrna_prober -i <span class="nv">$DATA_PATH</span>/sample1.zip <span class="nv">$DATA_PATH</span>/sample2.zip ... <span class="se">\</span>
                        -o <span class="nv">$DATA_PATH</span>/output_location/
</pre></div>
</div>
<p>An explanation of the <code class="xref py py-data docutils literal notranslate"><span class="pre">rrna_prober</span></code> submodule commands can be found in the following tables:</p>
<table border="1" class="colwidths-given docutils" id="id12">
<caption><span class="caption-text"><code class="xref py py-data docutils literal notranslate"><span class="pre">rrna_prober</span></code> required arguments</span><a class="headerlink" href="#id12" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="41%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Argument</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--input</span> <span class="pre">&lt;string&gt;</span> <span class="pre">[&lt;string&gt;</span> <span class="pre">...]</span></code></td>
<td>Space delimited list of zipped files (include full paths to these files)</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;string&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;string&gt;</span></code></td>
<td>Output file name to write output to</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="gene-dictionary-module">
<h2><code class="xref py py-data docutils literal notranslate"><span class="pre">gene_dictionary</span></code> module<a class="headerlink" href="#gene-dictionary-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="diffex-module">
<h2><code class="xref py py-data docutils literal notranslate"><span class="pre">diffex</span></code> module<a class="headerlink" href="#diffex-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="truncate-module">
<h2><code class="xref py py-data docutils literal notranslate"><span class="pre">truncate</span></code> module<a class="headerlink" href="#truncate-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="data-output">
<h2>Data Output<a class="headerlink" href="#data-output" title="Permalink to this headline">¶</a></h2>
<p>Running <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code> or <code class="xref py py-data docutils literal notranslate"><span class="pre">rnaseq</span></code> will output all intermediate and final data files in a tree structure as seen below:</p>
<a class="reference internal image-reference" href="content/ribopipe_overview.png"><img alt="content/ribopipe_overview.png" class="align-center" src="content/ribopipe_overview.png" style="width: 600px;" /></a>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="advanced-usage.html" class="btn btn-neutral float-right" title="Advanced Usage" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="quickstart.html" class="btn btn-neutral float-left" title="Quickstart" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>