

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Retrieving Data &mdash; xpresstools 0.0.1-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Reference File Modification" href="formatting.html" />
    <link rel="prev" title="General Usage" href="general-usage.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> xpresstools
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-usage.html">General Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Retrieving Data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#import-data-from-file">Import data from file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#import-metadata-from-file">Import metadata from file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#import-data-from-geo-dataset">Import data from GEO dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#catenate-raw-counts-files">Catenate raw counts files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-count-table-from-file-list">Create count table from file list</a></li>
<li class="toctree-l2"><a class="reference internal" href="#drop-samples">Drop samples</a></li>
<li class="toctree-l2"><a class="reference internal" href="#drop-label">Drop label</a></li>
<li class="toctree-l2"><a class="reference internal" href="#keep-labels">Keep labels</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rename-dataframe-column-names">Rename dataframe column names</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rename-genes-with-gtf">Rename genes with GTF</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rename-dataframe-row-names">Rename dataframe row names</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="formatting.html">Reference File Modification</a></li>
<li class="toctree-l1"><a class="reference internal" href="normalize.html">Data Normalization</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">xpresstools</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Retrieving Data</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/getting-data.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="retrieving-data">
<h1>Retrieving Data<a class="headerlink" href="#retrieving-data" title="Permalink to this headline">¶</a></h1>
<div class="section" id="import-data-from-file">
<h2>Import data from file<a class="headerlink" href="#import-data-from-file" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.get_df ( file_name, delimiter=’,’, low_memory=False, gene_axis=’row’ )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Get sequence dataframe from user-provided file.</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Dataset does not contain axis labels (i.e. a column header for ‘gene names’)</div>
<div class="line">- Dataset only has gene names and sample_ids as column headers and row indices. Orientation is flexible, but needs to be specified in options if genes are not rows</div>
<div class="line">- If orientation is not default, it is then specified or else function will not be able to properly format the dataframe for downstream application</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>file_name</strong>: Full path of file to import into pandas dataframe</div>
<div class="line"><strong>delimiter</strong>: Delimiter type for importing file, default: ‘,’</div>
<div class="line"><strong>low_memory</strong>: Specify memory limits for importing large files, default: False (allows for large imports)</div>
<div class="line"><strong>gene_axis</strong>: Orientiation of the data, where categorical data is either column-wise, (default: ‘col’) or row-wise (‘row’). Case insensitive</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data</strong>: Pandas dataframe with data matrix</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="o">&gt;</span> <span class="kn">import</span> <span class="nn">xpresstools</span> <span class="kn">as</span> <span class="nn">xp</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">get_df</span><span class="p">(</span><span class="s1">&#39;/path/to/data.csv&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="import-metadata-from-file">
<h2>Import metadata from file<a class="headerlink" href="#import-metadata-from-file" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.get_info ( file_name, delimiter=”,”, axis=”col”, sample_ids=0, labels=1 )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Get sample metadata from user-provided file</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line">* Data categories are not labeled</div>
<div class="line">* If orientation is not default, it is then specified or else function will not be able to properly format the dataframe for downstream application</div>
<div class="line"><br /></div>
<div class="line">Parameters:</div>
<div class="line"><strong>file_name</strong>: full path of file to import into pandas dataframe</div>
<div class="line"><strong>delimiter</strong>: delimiter type for importing file, default: ‘,’</div>
<div class="line"><strong>axis</strong>: Orientiation of the data, where categorical data is either column-wise, (default: ‘col’) or row-wise (‘row’). Case insensitive</div>
<div class="line"><strong>sample_ids</strong>: Column or row number where sample IDs are found (default: 0)</div>
<div class="line"><strong>labels</strong>: Column or row number where categorical label data are found (default: 1)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>metadata</strong>: Pandas dataframe with metadata</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="o">&gt;</span> <span class="kn">import</span> <span class="nn">xpresstools</span> <span class="kn">as</span> <span class="nn">xp</span>
<span class="o">&gt;</span> <span class="n">metadata</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">get_info</span><span class="p">(</span><span class="s1">&#39;/path/to/metadata.csv&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">metadata</span>
    <span class="mi">0</span>         <span class="mi">1</span>
<span class="mi">0</span>   <span class="n">GSM523242</span> <span class="n">mucosa_normal_colon</span>
<span class="mi">1</span>   <span class="n">GSM523243</span> <span class="n">mucosa_normal_colon</span>
<span class="mi">2</span>   <span class="n">GSM523244</span> <span class="n">mucosa_adenoma</span>
<span class="mi">3</span>   <span class="n">GSM523245</span> <span class="n">colonic_crypt_epithelial_cells_normal_colon</span>
<span class="o">...</span> <span class="o">...</span>       <span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="import-data-from-geo-dataset">
<h2>Import data from GEO dataset<a class="headerlink" href="#import-data-from-geo-dataset" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.get_geo ( geo_id, output_info=False )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Get sample data and metadata from a GEO database</div>
<div class="line"><br /></div>
<div class="line">Parameters:</div>
<div class="line"><strong>geo_id</strong>: GEO ID for dataset of interest, input is case insensitive (ex: GSE20716)</div>
<div class="line"><strong>output_info</strong>: Output long-form metadata to txt file if True (default: False)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data</strong>: Pandas dataframe with data matrix</div>
<div class="line"><strong>metadata</strong>: Pandas dataframe with metadata</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="o">&gt;</span> <span class="kn">import</span> <span class="nn">xpresstools</span> <span class="kn">as</span> <span class="nn">xp</span>
<span class="o">&gt;</span> <span class="n">data</span><span class="p">,</span> <span class="n">metadata</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">get_geo</span><span class="p">(</span><span class="s1">&#39;GSE20916&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">metadata</span>
    <span class="mi">0</span>         <span class="mi">1</span>
<span class="mi">0</span>   <span class="n">GSM523242</span> <span class="n">mucosa_normal_colon</span>
<span class="mi">1</span>   <span class="n">GSM523243</span> <span class="n">mucosa_normal_colon</span>
<span class="mi">2</span>   <span class="n">GSM523244</span> <span class="n">mucosa_adenoma</span>
<span class="mi">3</span>   <span class="n">GSM523245</span> <span class="n">colonic_crypt_epithelial_cells_normal_colon</span>
<span class="o">...</span> <span class="o">...</span>       <span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="catenate-raw-counts-files">
<h2>Catenate raw counts files<a class="headerlink" href="#catenate-raw-counts-files" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.catenate_files ( directory, file_suffix=’txt’, save_file=None, delimiter=’t’, drop_rows=0 )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Compiles expression counts from multiple files into one table. For example, HTSeq-count outputs each alignment file’s counts as a separate count file. This module will collect all single count files and compile them into a single count table.</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- File length of each is the same and ordered the same (same genes in the same order)</div>
<div class="line">- Files to parse are expected to be header-less and column[0] should be gene identifiers and column[1] should be expression values</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>directory</strong>: Path to directory containing raw counts files (only tested currently with HTSeq-count output files)</div>
<div class="line"><strong>file_suffix</strong>: Common suffix of all count files (default: ‘txt’). This feature is useful for modification if there other files in the directory that are not count files, as if they do not contain the same suffix, they will not be used in the function.</div>
<div class="line"><strong>save_file</strong>: Include if you want the resulting counts table saved for later use (default: None)</div>
<div class="line"><strong>delimiter</strong>: Delimiter style for expression files, will also output files if saved in this same format</div>
<div class="line"><strong>drop_rows</strong>: Number of rows to drop from the end of each count file. HTSeq-count provides 5 lines of summary statistics at the end of each file, so for HTSeq-count files, use drop_rows=5</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>count_table</strong>: Pandas dataframe with the catenated counts. Samples are along columns, genes are along rows</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">catenate_files</span><span class="p">(</span><span class="n">count_dir</span><span class="p">,</span> <span class="n">file_suffix</span><span class="o">=</span><span class="s1">&#39;counts.txt&#39;</span><span class="p">,</span> <span class="n">drop_rows</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">counts</span>
        <span class="n">S1_counts</span><span class="o">.</span><span class="n">txt</span> <span class="n">S2_counts</span><span class="o">.</span><span class="n">txt</span> <span class="n">S3_counts</span><span class="o">.</span><span class="n">txt</span> <span class="n">S4_counts</span><span class="o">.</span><span class="n">txt</span>
<span class="n">Gene1</span>   <span class="mi">66</span>            <span class="mi">59</span>            <span class="mi">1</span>             <span class="mi">82</span>
<span class="n">Gene2</span>   <span class="mi">35</span>            <span class="mi">0</span>             <span class="mi">7</span>             <span class="mi">72</span>
<span class="n">Gene3</span>   <span class="mi">20</span>            <span class="mi">70</span>            <span class="mi">87</span>            <span class="mi">78</span>
<span class="n">Gene4</span>   <span class="mi">96</span>            <span class="mi">7</span>             <span class="mi">93</span>            <span class="mi">38</span>
<span class="o">...</span>     <span class="o">...</span>           <span class="o">...</span>           <span class="o">...</span>           <span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="create-count-table-from-file-list">
<h2>Create count table from file list<a class="headerlink" href="#create-count-table-from-file-list" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.count_table ( file_list, gene_column=0, sample_column=1, sep=’t’ )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Collate HTseq counts files (similar to catenate_files(), but input is a file list)</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- No headers are included in the count files</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>file_list</strong>: List of files with the path names appended to each file to be collated into a single count table</div>
<div class="line"><strong>gene_column</strong>: Column location in all count files of gene names</div>
<div class="line"><strong>gene_column</strong>: Column location in all count files of samples</div>
<div class="line"><strong>sep</strong>: Separator of counts files</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>count_table</strong>: Pandas dataframe with the catenated counts. Samples are along columns, genes are along rows</div>
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="drop-samples">
<h2>Drop samples<a class="headerlink" href="#drop-samples" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.drop_samples ( data, ids )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Drop samples by sample IDs – pass in a list of names</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Dataframe axes have been properly formatted (samples are columns, genes are rows)</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Dataframe containing expression data</div>
<div class="line"><strong>ids</strong>: List of sample IDs to remove from the dataframe</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data</strong>: Pandas dataframe with modified data matrix</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">drop_samples</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">metadata</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;GSM523244&#39;</span><span class="p">])</span>
<span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="drop-label">
<h2>Drop label<a class="headerlink" href="#drop-label" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.drop_label ( data, info, label )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Drop samples by label group name</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Dataframe axes have been properly formatted (samples are columns, genes are rows)</div>
<div class="line">- Only one string is given to drop per call instance of function</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Dataframe containing expression data</div>
<div class="line"><strong>info</strong>: Dataframe containing sample information data</div>
<div class="line"><strong>label</strong>: Name of sample type to drop (string)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data</strong>: Pandas dataframe with modified data matrix</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">drop_label</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">metadata</span><span class="p">,</span> <span class="s1">&#39;mucosa_adenoma&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="keep-labels">
<h2>Keep labels<a class="headerlink" href="#keep-labels" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.keep_labels ( data, info, label_list=None )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Keep samples by list of label names</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Dataframe axes have been properly formatted (samples are columns, genes are rows)</div>
<div class="line">- Labels provided are in list format</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Dataframe containing expression data</div>
<div class="line"><strong>info</strong>: Dataframe containing sample information data</div>
<div class="line"><strong>labels</strong>: List of sample types to keep</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data</strong>: Pandas dataframe with modified data matrix</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">keep_labels</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">metadata</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;mucosa_normal_colon&#39;</span><span class="p">,</span> <span class="s1">&#39;mucosa_adenoma&#39;</span><span class="p">])</span>
<span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="rename-dataframe-column-names">
<h2>Rename dataframe column names<a class="headerlink" href="#rename-dataframe-column-names" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.rename_cols ( data, converters )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Rename column names using dataframe</div>
<div class="line"><br /></div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Dataframe to rename column names</div>
<div class="line"><strong>converters</strong>: Dataframe where column 0 contains old names and column 1 contains new names</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data</strong>: Pandas dataframe with modified data matrix</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">conversion_table</span>
    <span class="mi">0</span>         <span class="mi">1</span>
<span class="mi">0</span>   <span class="n">GSM523242</span> <span class="n">normal</span>
<span class="mi">1</span>   <span class="n">GSM523244</span> <span class="n">adenoma</span>
<span class="mi">2</span>   <span class="n">GSM523245</span> <span class="n">normal</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">rename_cols</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">conversion_table</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">normal</span>    <span class="n">GSM523243</span> <span class="n">adenoma</span>   <span class="n">normal</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="rename-genes-with-gtf">
<h2>Rename genes with GTF<a class="headerlink" href="#rename-genes-with-gtf" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.convert_names_gtf ( data, gtf, orig_name_label=’gene_id “’, orig_name_location=0, new_name_label=’gene_name “’, new_name_location=1, refill=None, sep=’t’ )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Convert row names (genes) of dataframe using GTF as reference for new name</div>
<div class="line"><br /></div>
<div class="line">Important Notes:</div>
<div class="line-block">
<div class="line">- A cursory look at the GTF may be required to determine where in the final field the conversion data lies. Position is relative to delimiter in the final field (usally a “;”), so if the new name is in the third position, new_name_location=2, etc.</div>
<div class="line">- This function is pulling original and new gene name information from any row where the third field is “gene”. You can run <code class="xref py py-data docutils literal notranslate"><span class="pre">cat</span> <span class="pre">transcripts.gtf</span> <span class="pre">|</span> <span class="pre">awk</span> <span class="pre">'$3</span> <span class="pre">==</span> <span class="pre">&quot;gene&quot;'</span> <span class="pre">|</span> <span class="pre">less</span> <span class="pre">-S</span></code> from the command line of your reference file to identify the positions of the required text fields</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Dataframe to convert rows names</div>
<div class="line"><strong>gtf</strong>: Path and name of gtf reference file</div>
<div class="line"><strong>orig_name_label</strong>: Label of original name (usually a “gene_id “’)</div>
<div class="line"><strong>orig_name_location</strong>: Position in last column of GTF where relevant data is found (i.e. 0 would be the first sub-string before the first comma, 3 would be the third sub-string after the second comma before the third comma)</div>
<div class="line"><strong>new_name_label</strong>: Label of original name (usually “gene_name “)</div>
<div class="line"><strong>new_name_location</strong>: Position in last column of GTF where relevant data is found (i.e. 0 would be the first sub-string before the first comma, 3 would be the third sub-string after the second comma before the third comma)</div>
<div class="line"><strong>refill</strong>: In some cases, where common gene names are unavailable, the dataframe will fill the gene name with the improper field of the GTF. In this case, specify this improper string and these values will be replaced with the original name</div>
<div class="line"><strong>sep</strong>: GTF delimiter (usually tab-delimited)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data</strong>: Pandas dataframe with modified data matrix</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
     <span class="n">gene_names</span>     <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">0</span>    <span class="n">YXZ1034C</span>       <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1</span>    <span class="n">YXA7834D</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">2</span>    <span class="n">YXZ349C</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">3</span>    <span class="n">YXZ1994A</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>  <span class="o">...</span>            <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">convert_names_gtf</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;/path/to/transcripts.gtf&#39;</span><span class="p">,</span> <span class="n">new_name_label</span><span class="o">=</span><span class="s1">&#39;gene_name </span><span class="se">\&quot;</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">new_name_location</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
     <span class="n">gene_names</span>    <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">0</span>    <span class="n">Gene1</span>         <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1</span>    <span class="n">Gene2</span>         <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">2</span>    <span class="n">Gene3</span>         <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">3</span>    <span class="n">Gene4</span>         <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>  <span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="rename-dataframe-row-names">
<h2>Rename dataframe row names<a class="headerlink" href="#rename-dataframe-row-names" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.rename_rows ( data, converters, label=’index’ )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Rename values in an index (row names) or a column</div>
<div class="line"><br /></div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Dataframe to rename rows of a column</div>
<div class="line"><strong>converters</strong>: Dataframe where column 0 contains old names and column 1 contains new names</div>
<div class="line"><strong>label</strong>: Name of column to convert names; if ‘index’ is provided, will rename the index of the dataframe</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data</strong>: Pandas dataframe with modified data matrix</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">conversion_table</span>
    <span class="mi">0</span>         <span class="mi">1</span>
<span class="mi">0</span>   <span class="mi">1007</span><span class="n">_s_at</span> <span class="n">Gene1</span>
<span class="mi">1</span>   <span class="mi">121</span><span class="n">_at</span> <span class="n">Gene2</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">rename_rows</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">conversion_table</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
              <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="n">Gene1</span>         <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="n">Gene2</span>         <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
     <span class="n">gene_names</span>    <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">0</span>    <span class="mi">1007</span><span class="n">_s_at</span>     <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1</span>    <span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">2</span>    <span class="mi">121</span><span class="n">_at</span>        <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">3</span>    <span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>  <span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">conversion_table</span>
    <span class="mi">0</span>         <span class="mi">1</span>
<span class="mi">0</span>   <span class="mi">1007</span><span class="n">_s_at</span> <span class="n">Gene1</span>
<span class="mi">1</span>   <span class="mi">121</span><span class="n">_at</span> <span class="n">Gene2</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">rename_rows</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">conversion_table</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;gene_names&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
     <span class="n">gene_names</span>    <span class="n">GSM523242</span> <span class="n">GSM523243</span> <span class="n">GSM523244</span> <span class="n">GSM523245</span> <span class="o">...</span>
<span class="mi">0</span>    <span class="n">Gene1</span>         <span class="mf">8.98104</span>   <span class="mf">8.59941</span>   <span class="mf">8.25395</span>   <span class="mf">8.72981</span>   <span class="o">...</span>
<span class="mi">1</span>    <span class="mi">1053</span><span class="n">_at</span>       <span class="mf">5.84313</span>   <span class="mf">6.59168</span>   <span class="mf">8.27881</span>   <span class="mf">6.64005</span>   <span class="o">...</span>
<span class="mi">2</span>    <span class="n">Gene2</span>         <span class="mf">6.17189</span>   <span class="mf">5.73603</span>   <span class="mf">5.55673</span>   <span class="mf">5.69374</span>   <span class="o">...</span>
<span class="mi">3</span>    <span class="mi">1294</span><span class="n">_at</span>       <span class="mf">6.97009</span>   <span class="mf">6.80003</span>   <span class="mf">5.56620</span>   <span class="mf">7.43816</span>   <span class="o">...</span>
<span class="o">...</span>  <span class="o">...</span>           <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>       <span class="o">...</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="formatting.html" class="btn btn-neutral float-right" title="Reference File Modification" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="general-usage.html" class="btn btn-neutral float-left" title="General Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>