

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Normalization and Quality Control &mdash; xpresstools 0.0.1-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Analysis" href="analysis.html" />
    <link rel="prev" title="Retrieving Data" href="getting-data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> xpresstools
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">XPRESStools</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-usage.html">General Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-data.html">Retrieving Data</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Normalization and Quality Control</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#rpm-reads-per-million">RPM (Reads per Million)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#r-fpkm-reads-fragments-per-kilobase-million">R/FPKM (Reads/Fragments per Kilobase Million)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#te-translation-efficiency">TE (Translation Efficiency)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#log-transformation">Log transformation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#batch-normalize">Batch Normalize</a></li>
<li class="toctree-l2"><a class="reference internal" href="#clean-dataset">Clean dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#set-gene-threshold">Set Gene Threshold</a></li>
<li class="toctree-l2"><a class="reference internal" href="#prepare-xpresstools-dataset">Prepare XPRESStools dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#check-sample-expression-distributions">Check Sample Expression Distributions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#microarray-probe-collapse">Microarray Probe Collapse</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference-building.html">Reference Building</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">xpresstools</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Normalization and Quality Control</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/normalize.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="normalization-and-quality-control">
<h1>Normalization and Quality Control<a class="headerlink" href="#normalization-and-quality-control" title="Permalink to this headline">¶</a></h1>
<div class="section" id="rpm-reads-per-million">
<h2>RPM (Reads per Million)<a class="headerlink" href="#rpm-reads-per-million" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.rpm ( data )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Perform reads per million sample normalization on RNAseq data</div>
<div class="line"><br /></div>
<div class="line">Formula:</div>
<div class="line"><span class="math notranslate nohighlight">\(RPM = \frac{\#\ number\ reads\ per\ gene\ x\ 1e6}{\#\ mapped\ reads\ per\ sample}\)</span></div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Dataframe contains raw count data, where samples are along columns and genes across rows</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe with counts values</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_rpm</strong>: Pandas dataframe with RPM-normalized</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
      <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="n">Gene1</span> <span class="mi">66</span>          <span class="mi">59</span>          <span class="mi">1</span>           <span class="mi">82</span>          <span class="mi">45</span>
<span class="n">Gene2</span> <span class="mi">35</span>          <span class="mi">0</span>           <span class="mi">7</span>           <span class="mi">72</span>          <span class="mi">2</span>
<span class="n">Gene3</span> <span class="mi">20</span>          <span class="mi">70</span>          <span class="mi">85</span>          <span class="mi">78</span>          <span class="mi">36</span>
<span class="n">Gene4</span> <span class="mi">96</span>          <span class="mi">7</span>           <span class="mi">93</span>          <span class="mi">38</span>          <span class="mi">85</span>
<span class="n">Gene5</span> <span class="mi">73</span>          <span class="mi">41</span>          <span class="mi">92</span>          <span class="mi">77</span>          <span class="mi">26</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">rpm</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
      <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="n">Gene1</span> <span class="mf">227586.2069</span> <span class="mf">333333.3333</span> <span class="mf">3597.1223</span>   <span class="mf">236311.2392</span> <span class="mf">231958.7629</span>
<span class="n">Gene2</span> <span class="mf">120689.6552</span> <span class="mf">0.0000</span>      <span class="mf">25179.8561</span>  <span class="mf">207492.7954</span> <span class="mf">10309.2784</span>
<span class="n">Gene3</span> <span class="mf">68965.5172</span>  <span class="mf">395480.2260</span> <span class="mf">305755.3957</span> <span class="mf">224783.8617</span> <span class="mf">185567.0103</span>
<span class="n">Gene4</span> <span class="mf">331034.4828</span> <span class="mf">39548.0226</span>  <span class="mf">334532.3741</span> <span class="mf">109510.0865</span> <span class="mf">438144.3299</span>
<span class="n">Gene5</span> <span class="mf">251724.1379</span> <span class="mf">231638.4181</span> <span class="mf">330935.2518</span> <span class="mf">221902.0173</span> <span class="mf">134020.6186</span>
</pre></div>
</div>
</div>
<div class="section" id="r-fpkm-reads-fragments-per-kilobase-million">
<h2>R/FPKM (Reads/Fragments per Kilobase Million)<a class="headerlink" href="#r-fpkm-reads-fragments-per-kilobase-million" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.r_fpkm ( data, gtf, gene_name_prefix=’gene_id “’, gene_name_location=0, sep=’t’ )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Perform reads/fragments per kilobase million sample normalization on RNAseq data</div>
<div class="line"><br /></div>
<div class="line">Formulae:</div>
<div class="line"><span class="math notranslate nohighlight">\(RPKM = \frac{\#\ number\ reads\ per\ gene\ x\ 1e6\ x\ 1e3}{\#\ mapped\ reads\ per\ sample\ x\ gene\ length\ (bp)}\)</span></div>
<div class="line"><br /></div>
<div class="line"><span class="math notranslate nohighlight">\(FPKM = \frac{\#\ number\ fragments\ per\ gene\ x\ 1e6\ x\ 1e3}{\#\ mapped\ fragments\ per\ sample\ x\ gene\ length\ (bp)}\)</span></div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Dataframe contains raw count data, where samples are along columns and genes across rows</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe with counts values</div>
<div class="line"><strong>gtf</strong>: GTF reference file path and name</div>
<div class="line"><strong>gene_name_prefix</strong>: Label of original name (usually “gene_name “), must match the gene naming type in the input dataframe</div>
<div class="line"><strong>gene_name_location</strong>: Position in last column of GTF where relevant data is found (i.e. 0 would be the first sub-string before the first comma, 3 would be the third sub-string after the second comma before the third comma)</div>
<div class="line"><strong>sep</strong>: GTF delimiter (usually tab-delimited)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_rpkm</strong>: Pandas dataframe with R/FPKM-normalized</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
      <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="n">Gene1</span> <span class="mi">66</span>          <span class="mi">59</span>          <span class="mi">1</span>           <span class="mi">82</span>          <span class="mi">45</span>
<span class="n">Gene2</span> <span class="mi">35</span>          <span class="mi">0</span>           <span class="mi">7</span>           <span class="mi">72</span>          <span class="mi">2</span>
<span class="n">Gene3</span> <span class="mi">20</span>          <span class="mi">70</span>          <span class="mi">85</span>          <span class="mi">78</span>          <span class="mi">36</span>
<span class="n">Gene4</span> <span class="mi">96</span>          <span class="mi">7</span>           <span class="mi">93</span>          <span class="mi">38</span>          <span class="mi">85</span>
<span class="n">Gene5</span> <span class="mi">73</span>          <span class="mi">41</span>          <span class="mi">92</span>          <span class="mi">77</span>          <span class="mi">26</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">r_fpkm</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;/path/to/transcripts.gtf&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
      <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="n">Gene1</span> <span class="mf">15006.3436</span>  <span class="mf">21978.9881</span>  <span class="mf">237.1833</span>    <span class="mf">15581.6457</span>  <span class="mf">15294.6567</span>
<span class="n">Gene2</span> <span class="mf">18516.3632</span>  <span class="mf">0.0000</span>      <span class="mf">3863.1261</span>   <span class="mf">31833.8133</span>  <span class="mf">1581.6628</span>
<span class="n">Gene3</span> <span class="mf">1552.2985</span>   <span class="mf">8901.5987</span>   <span class="mf">6882.0428</span>   <span class="mf">5059.5089</span>   <span class="mf">4176.8031</span>
<span class="n">Gene4</span> <span class="mf">10220.8992</span>  <span class="mf">1221.0702</span>   <span class="mf">10328.8988</span>  <span class="mf">3381.1932</span>   <span class="mf">13527.9835</span>
<span class="n">Gene5</span> <span class="mf">12188.2602</span>  <span class="mf">11215.7274</span>  <span class="mf">16023.5923</span>  <span class="mf">10744.2995</span>  <span class="mf">6489.1599</span>
</pre></div>
</div>
</div>
<div class="section" id="te-translation-efficiency">
<h2>TE (Translation Efficiency)<a class="headerlink" href="#te-translation-efficiency" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.te ( data, samples=None, log2=True )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Transform paired ribosome footprint and RNA samples to translation efficiency measures</div>
<div class="line"><br /></div>
<div class="line">Formulae:</div>
<div class="line"><span class="math notranslate nohighlight">\(log_2TE = log_2(\frac{RPF\ reads\ per\ gene}{total\ RNA\ reads\ per\ gene})\)</span></div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Dataframe contains normalized count data, where samples are along columns and genes across rows (unless otherwise desired)</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe with counts values</div>
<div class="line"><strong>samples</strong>: GTF reference file path and name</div>
<div class="line"><strong>log2</strong>: Label of original name (usually “gene_name “), must match the gene naming type in the input dataframe</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_te</strong>: Pandas dataframe TE transformed</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000227232</span> <span class="mf">66.34</span>   <span class="mf">59.13</span>   <span class="mf">1.90</span>    <span class="mf">82.49</span>
<span class="n">ENSG00000240361</span> <span class="mf">35.73</span>   <span class="mf">0.00</span>    <span class="mf">7.38</span>    <span class="mf">72.94</span>
<span class="n">ENSG00000238009</span> <span class="mf">20.02</span>   <span class="mf">70.21</span>   <span class="mf">85.10</span>   <span class="mf">78.87</span>
<span class="n">ENSG00000241860</span> <span class="mf">96.23</span>   <span class="mf">7.49</span>    <span class="mf">93.49</span>   <span class="mf">38.39</span>
<span class="n">ENSG00000187634</span> <span class="mf">73.91</span>   <span class="mf">41.28</span>   <span class="mf">92.27</span>   <span class="mf">77.93</span>
<span class="o">&gt;</span> <span class="n">te_data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">te</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample1&#39;</span><span class="p">,</span><span class="s1">&#39;sample2&#39;</span><span class="p">])</span>
<span class="o">&gt;</span> <span class="n">te_data</span>
                <span class="n">sample1</span> <span class="n">sample2</span>
<span class="n">ENSG00000227232</span> <span class="mf">0.1657</span>  <span class="o">-</span><span class="mf">5.3679</span>
<span class="n">ENSG00000240361</span> <span class="mf">8.4850</span>  <span class="o">-</span><span class="mf">3.2876</span>
<span class="n">ENSG00000238009</span> <span class="o">-</span><span class="mf">1.8051</span> <span class="mf">0.1095</span>
<span class="n">ENSG00000241860</span> <span class="mf">3.6658</span>  <span class="mf">1.2819</span>
<span class="n">ENSG00000187634</span> <span class="mf">0.8388</span>  <span class="mf">0.2434</span>
<span class="o">&gt;</span> <span class="n">te_data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">te</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">te_data</span>
                <span class="n">s1_rpf_te</span> <span class="n">s2_rpf_te</span>
<span class="n">ENSG00000227232</span> <span class="mf">0.1657</span>    <span class="o">-</span><span class="mf">5.3679</span>
<span class="n">ENSG00000240361</span> <span class="mf">8.4850</span>    <span class="o">-</span><span class="mf">3.2876</span>
<span class="n">ENSG00000238009</span> <span class="o">-</span><span class="mf">1.8051</span>   <span class="mf">0.1095</span>
<span class="n">ENSG00000241860</span> <span class="mf">3.6658</span>    <span class="mf">1.2819</span>
<span class="n">ENSG00000187634</span> <span class="mf">0.8388</span>    <span class="mf">0.2434</span>
</pre></div>
</div>
</div>
<div class="section" id="log-transformation">
<h2>Log transformation<a class="headerlink" href="#log-transformation" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.log_scale ( data, log_base=10 )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Log-scale a sample-normalized dataframe</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for XPRESStools usage</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe with counts values</div>
<div class="line"><strong>log_base</strong>: Log base to use for transformation (default: 10; or 2)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_log</strong>: Pandas dataframe log-scaled</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000227232</span> <span class="mf">66.34</span>   <span class="mf">59.13</span>   <span class="mf">1.90</span>    <span class="mf">82.49</span>
<span class="n">ENSG00000240361</span> <span class="mf">35.73</span>   <span class="mf">0.00</span>    <span class="mf">7.38</span>    <span class="mf">72.94</span>
<span class="n">ENSG00000238009</span> <span class="mf">20.02</span>   <span class="mf">70.21</span>   <span class="mf">85.10</span>   <span class="mf">78.87</span>
<span class="n">ENSG00000241860</span> <span class="mf">96.23</span>   <span class="mf">7.49</span>    <span class="mf">93.49</span>   <span class="mf">38.39</span>
<span class="n">ENSG00000187634</span> <span class="mf">73.91</span>   <span class="mf">41.28</span>   <span class="mf">92.27</span>   <span class="mf">77.93</span>
<span class="o">&gt;</span> <span class="n">log_data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">log_scale</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">log_data</span>
                <span class="n">s1_rpf</span>    <span class="n">s1_rna</span>    <span class="n">s2_rpf</span>    <span class="n">s2_rna</span>
<span class="n">ENSG00000227232</span> <span class="mf">1.822430</span>  <span class="mf">1.772542</span>  <span class="mf">0.301030</span>  <span class="mf">1.916927</span>
<span class="n">ENSG00000240361</span> <span class="mf">1.554247</span>  <span class="o">-</span><span class="mf">1.000000</span> <span class="mf">0.873902</span>  <span class="mf">1.863561</span>
<span class="n">ENSG00000238009</span> <span class="mf">1.303628</span>  <span class="mf">1.847017</span>  <span class="mf">1.930440</span>  <span class="mf">1.897462</span>
<span class="n">ENSG00000241860</span> <span class="mf">1.983762</span>  <span class="mf">0.880242</span>  <span class="mf">1.971229</span>  <span class="mf">1.585348</span>
<span class="n">ENSG00000187634</span> <span class="mf">1.869290</span>  <span class="mf">1.616790</span>  <span class="mf">1.965531</span>  <span class="mf">1.892262</span>
</pre></div>
</div>
</div>
<div class="section" id="batch-normalize">
<h2>Batch Normalize<a class="headerlink" href="#batch-normalize" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.batch_normalize ( input_file, batch_file, output_file, input_sep=’,’, batch_sep=’,’ )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Control for batch effects between datasets</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for XPRESStools usage where samples are normalized previously if desired</div>
<div class="line">- Requires a properly formatted dataframe complying to SVA COMBAT info file (see example below)</div>
<div class="line">- R is installed on your machine and is in your $PATH</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>input_file</strong>: Input dataframe file with values (can be normalized or unnormalized)</div>
<div class="line"><strong>batch_file</strong>: Input dataframe containing batch effect information, column naming convention must be followed and <em>is</em> case-sensitive</div>
<div class="line"><strong>output_file</strong>: Output path and file name for batch normalized data</div>
<div class="line"><strong>input_sep</strong>: Delimiter for input_file</div>
<div class="line"><strong>batch_sep</strong>: Delimiter for batch_file</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/expression.csv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000227232</span> <span class="mf">66.34</span>   <span class="mf">59.13</span>   <span class="mf">1.90</span>    <span class="mf">82.49</span>
<span class="n">ENSG00000240361</span> <span class="mf">35.73</span>   <span class="mf">0.00</span>    <span class="mf">7.38</span>    <span class="mf">72.94</span>
<span class="n">ENSG00000238009</span> <span class="mf">20.02</span>   <span class="mf">70.21</span>   <span class="mf">85.10</span>   <span class="mf">78.87</span>
<span class="n">ENSG00000241860</span> <span class="mf">96.23</span>   <span class="mf">7.49</span>    <span class="mf">93.49</span>   <span class="mf">38.39</span>
<span class="n">ENSG00000187634</span> <span class="mf">73.91</span>   <span class="mf">41.28</span>   <span class="mf">92.27</span>   <span class="mf">77.93</span>
<span class="o">&gt;</span> <span class="n">batch</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/batch_info.csv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">batch</span>
  <span class="n">Sample</span>  <span class="n">Batch</span>
<span class="mi">0</span> <span class="n">s1_rpf</span>  <span class="n">batch1</span>
<span class="mi">1</span> <span class="n">s1_rna</span>  <span class="n">batch2</span>
<span class="mi">2</span> <span class="n">s2_rpf</span>  <span class="n">batch1</span>
<span class="mi">3</span> <span class="n">s2_rna</span>  <span class="n">batch2</span>
<span class="o">&gt;</span> <span class="n">xp</span><span class="o">.</span><span class="n">batch_normalize</span><span class="p">(</span><span class="s1">&#39;/path/to/expression.csv&#39;</span><span class="p">,</span> <span class="s1">&#39;/path/to/batch_info.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="clean-dataset">
<h2>Clean dataset<a class="headerlink" href="#clean-dataset" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.clean_df ( data, axis=0 )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Cleans NULL values from axis and clears duplicate indices</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for XPRESStools usage</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe file with values (can be normalized or unnormalized)</div>
<div class="line"><strong>axis</strong>: Axis to clean NaN values from (default: 0, which corresponds to rows)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_clean</strong>: Cleaned pandas dataframe</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000227232</span> <span class="mf">66.34</span>   <span class="mf">59.13</span>   <span class="mf">1.90</span>    <span class="n">NA</span>
<span class="n">ENSG00000240361</span> <span class="mf">35.73</span>   <span class="mf">0.00</span>    <span class="mf">7.38</span>    <span class="mf">72.94</span>
<span class="n">Gene2</span>           <span class="mf">20.02</span>   <span class="mf">70.21</span>   <span class="mf">85.10</span>   <span class="mf">78.87</span>
<span class="n">Gene2</span>           <span class="mf">96.23</span>   <span class="mf">7.49</span>    <span class="mf">93.49</span>   <span class="mf">38.39</span>
<span class="n">ENSG00000187634</span> <span class="mf">73.91</span>   <span class="n">NA</span>      <span class="mf">92.27</span>   <span class="mf">77.93</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">clean_df</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000240361</span> <span class="mf">35.73</span>   <span class="mf">0.00</span>    <span class="mf">7.38</span>    <span class="mf">72.94</span>
</pre></div>
</div>
</div>
<div class="section" id="set-gene-threshold">
<h2>Set Gene Threshold<a class="headerlink" href="#set-gene-threshold" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.threshold ( data, minimum=None, maximum=None )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Cleans gene axis (assumed to by rows) of genes containing values below or above user-determined thresholds</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for XPRESStools usage</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe file with values (can be normalized or unnormalized)</div>
<div class="line"><strong>minimum</strong>: Minimum value all samples need of a given gene to avoid dropping across all samples</div>
<div class="line"><strong>maximum</strong>: Maximum value all samples can have of a given gene to avoid dropping across all samples</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_clean</strong>: Cleaned pandas dataframe</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000227232</span> <span class="mf">66.34</span>   <span class="mf">59.13</span>   <span class="mf">1.90</span>    <span class="mf">82.49</span>
<span class="n">ENSG00000240361</span> <span class="mf">35.73</span>   <span class="mf">0.00</span>    <span class="mf">7.38</span>    <span class="mf">72.94</span>
<span class="n">ENSG00000238009</span> <span class="mf">20.02</span>   <span class="mf">70.21</span>   <span class="mf">85.10</span>   <span class="mf">78.87</span>
<span class="n">ENSG00000241860</span> <span class="mf">96.23</span>   <span class="mf">7.49</span>    <span class="mf">93.49</span>   <span class="mf">38.39</span>
<span class="n">ENSG00000187634</span> <span class="mf">73.91</span>   <span class="mf">41.28</span>   <span class="mf">92.27</span>   <span class="mf">77.93</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">threshold</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">minimum</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000238009</span> <span class="mf">20.02</span>   <span class="mf">70.21</span>   <span class="mf">85.10</span>   <span class="mf">78.87</span>
<span class="n">ENSG00000241860</span> <span class="mf">96.23</span>   <span class="mf">7.49</span>    <span class="mf">93.49</span>   <span class="mf">38.39</span>
<span class="n">ENSG00000187634</span> <span class="mf">73.91</span>   <span class="mf">41.28</span>   <span class="mf">92.27</span>   <span class="mf">77.93</span>
</pre></div>
</div>
</div>
<div class="section" id="prepare-xpresstools-dataset">
<h2>Prepare XPRESStools dataset<a class="headerlink" href="#prepare-xpresstools-dataset" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.prep_data ( data, info, gene_scale=True, print_means=False )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Prepare dataframe for downstream analyses</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for XPRESStools usage (genes as rows, samples as columns)</div>
<div class="line">- Requires properly formatted XPRESStools metadata dataframe</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: XPRESStools formatted dataframe of expression values</div>
<div class="line"><strong>info</strong>: XPRESStools formatted sample info dataframe</div>
<div class="line"><strong>gene_scale</strong>: Scale genes (rows) of data</div>
<div class="line"><strong>print_means</strong>: Print means for each sample verification</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_normalized</strong>: Normalized pandas dataframe</div>
<div class="line"><strong>data_labeled</strong>: Labeled pandas dataframe</div>
</div>
</div>
<div class="section" id="check-sample-expression-distributions">
<h2>Check Sample Expression Distributions<a class="headerlink" href="#check-sample-expression-distributions" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.check_samples ( data )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Visualize gene expression distributions on a sample-by-sample basis</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for XPRESStools usage</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe file with values (can be normalized or unnormalized)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line">Boxplot with samples on the x-axis and lump expression distributions for all genes in that sample</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">xp</span><span class="o">.</span><span class="n">check_samples</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../_images/quality.png"><img alt="../_images/quality.png" class="align-left" src="../_images/quality.png" style="width: 576.0px; height: 288.0px;" /></a>
</div>
<div class="section" id="microarray-probe-collapse">
<h2>Microarray Probe Collapse<a class="headerlink" href="#microarray-probe-collapse" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpresstools.probe_collapse ( data, reference, gene_list=None, no_multimappers=True )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Remove multimapping probes and collapse probes mapping to the same gene by averaging the values for those probes per sample</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for XPRESStools usage</div>
<div class="line">- Assumes GPL .txt file from NCBI is tab delimited</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe file with values (can be normalized or unnormalized)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_collapsed</strong>: Pandas dataframe file probes collapsed and the corresponding gene names listed</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="analysis.html" class="btn btn-neutral float-right" title="Analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="getting-data.html" class="btn btn-neutral float-left" title="Retrieving Data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>