

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Normalization and Quality Control &mdash; XPRESSplot 0.2.0-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Analysis" href="analysis.html" />
    <link rel="prev" title="Retrieving Data" href="getting-data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> XPRESSplot
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">XPRESSplot</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Wait! I’ve never programmed before!</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-usage.html">General Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-data.html">Retrieving Data</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Normalization and Quality Control</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#rpm-reads-per-million">RPM (Reads per Million)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#r-fpkm-reads-fragments-per-kilobase-million-per-million-mapped-reads">R/FPKM (Reads/Fragments per Kilobase Million per Million Mapped Reads)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#tpm-transcripts-per-million">TPM (Transcripts per Million)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#batch-normalize">Batch Normalize</a></li>
<li class="toctree-l2"><a class="reference internal" href="#clean-data">Clean Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#set-gene-threshold">Set Gene Threshold</a></li>
<li class="toctree-l2"><a class="reference internal" href="#prepare-xpressplot-dataset">Prepare xpressplot Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#check-sample-expression-distributions">Check Sample Expression Distributions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#microarray-probe-collapse">Microarray Probe Collapse</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="analysis.html">Analysis</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">XPRESSplot</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Normalization and Quality Control</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/normalize.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="normalization-and-quality-control">
<h1>Normalization and Quality Control<a class="headerlink" href="#normalization-and-quality-control" title="Permalink to this headline">¶</a></h1>
<div class="section" id="rpm-reads-per-million">
<h2>RPM (Reads per Million)<a class="headerlink" href="#rpm-reads-per-million" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpressplot.rpm ( data )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Perform reads per million sample normalization on RNAseq data</div>
<div class="line"><br /></div>
<div class="line">Formula:</div>
<div class="line"><span class="math notranslate nohighlight">\(RPM_{g} = \frac{1e6 \cdot r_{\textit{ge}}}{\sum_{g=1}^{n} r_{\textit{ge}}}\)</span></div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Dataframe contains raw count data, where samples are along columns and genes across rows</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe with counts values</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_rpm</strong>: Pandas dataframe with RPM-normalized</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
      <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="n">Gene1</span> <span class="mi">66</span>          <span class="mi">59</span>          <span class="mi">1</span>           <span class="mi">82</span>          <span class="mi">45</span>
<span class="n">Gene2</span> <span class="mi">35</span>          <span class="mi">0</span>           <span class="mi">7</span>           <span class="mi">72</span>          <span class="mi">2</span>
<span class="n">Gene3</span> <span class="mi">20</span>          <span class="mi">70</span>          <span class="mi">85</span>          <span class="mi">78</span>          <span class="mi">36</span>
<span class="n">Gene4</span> <span class="mi">96</span>          <span class="mi">7</span>           <span class="mi">93</span>          <span class="mi">38</span>          <span class="mi">85</span>
<span class="n">Gene5</span> <span class="mi">73</span>          <span class="mi">41</span>          <span class="mi">92</span>          <span class="mi">77</span>          <span class="mi">26</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">rpm</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
      <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="n">Gene1</span> <span class="mf">227586.2069</span> <span class="mf">333333.3333</span> <span class="mf">3597.1223</span>   <span class="mf">236311.2392</span> <span class="mf">231958.7629</span>
<span class="n">Gene2</span> <span class="mf">120689.6552</span> <span class="mf">0.0000</span>      <span class="mf">25179.8561</span>  <span class="mf">207492.7954</span> <span class="mf">10309.2784</span>
<span class="n">Gene3</span> <span class="mf">68965.5172</span>  <span class="mf">395480.2260</span> <span class="mf">305755.3957</span> <span class="mf">224783.8617</span> <span class="mf">185567.0103</span>
<span class="n">Gene4</span> <span class="mf">331034.4828</span> <span class="mf">39548.0226</span>  <span class="mf">334532.3741</span> <span class="mf">109510.0865</span> <span class="mf">438144.3299</span>
<span class="n">Gene5</span> <span class="mf">251724.1379</span> <span class="mf">231638.4181</span> <span class="mf">330935.2518</span> <span class="mf">221902.0173</span> <span class="mf">134020.6186</span>
</pre></div>
</div>
</div>
<div class="section" id="r-fpkm-reads-fragments-per-kilobase-million-per-million-mapped-reads">
<h2>R/FPKM (Reads/Fragments per Kilobase Million per Million Mapped Reads)<a class="headerlink" href="#r-fpkm-reads-fragments-per-kilobase-million-per-million-mapped-reads" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpressplot.r_fpkm ( data, gtf, feature_type=’exon’, identifier=’gene_name’, sep=’t’ )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Perform reads/fragments per kilobase per million mapped reads sample normalization on RNAseq data</div>
<div class="line"><br /></div>
<div class="line">Formulae:</div>
<div class="line"><span class="math notranslate nohighlight">\(RPKM_{g} = \frac{1e9 \cdot r_{\textit{ge}}}{(\sum_{g=1}^{n} r_{\textit{ge}}) \cdot \textit{l} _{\textit{ge}}}\)</span></div>
<div class="line"><br /></div>
<div class="line"><span class="math notranslate nohighlight">\(FPKM_{g} = \frac{1e9 \cdot f_{\textit{ge}}}{(\sum_{g=1}^{n} f_{\textit{ge}}) \cdot \textit{l} _{\textit{ge}}}\)</span></div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Dataframe contains raw count data, where samples are along columns and genes across rows</div>
<div class="line">- As FPKM was developed for paired-end sequencing, it accounts for two reads being able to map to one fragment. Therefore, input counts should have accounted for this is the counting step of sequence quantification. If specifying a paired-end alignment in XPRESSpipe, this will have been accounted for.</div>
<div class="line">- By default, this will take the longest transcript based on combined exon length. If you prefer to use the cumulative exon length of the Ensembl canonical transcript, you must first curate the GTF using <code class="xref py py-data docutils literal notranslate"><span class="pre">xpresspipe</span> <span class="pre">modifyGTF</span> <span class="pre">-g</span> <span class="pre">transcripts.gtf</span> <span class="pre">-l</span></code></div>
<div class="line">- If you performed isoform quantification and wish to length normalize each isoform, provide <code class="xref py py-data docutils literal notranslate"><span class="pre">identifier='transcript_id'</span></code></div>
<div class="line">- If you would like to normalize based on CDS length, provide <code class="xref py py-data docutils literal notranslate"><span class="pre">feature_type='CDS'</span></code></div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe with counts values</div>
<div class="line"><strong>gtf</strong>: GTF reference file path and name</div>
<div class="line"><strong>feature_type</strong>: Label of feature to use in length normalization</div>
<div class="line"><strong>identifier</strong>: Label for how to group genes/transcripts. If normalizing for a gene, use <code class="xref py py-data docutils literal notranslate"><span class="pre">gene_name</span></code> or <code class="xref py py-data docutils literal notranslate"><span class="pre">gene_id</span></code>. If performing length normalization for isoforms, provide <code class="xref py py-data docutils literal notranslate"><span class="pre">transcript_id</span></code></div>
<div class="line"><strong>sep</strong>: GTF delimiter (usually tab-delimited)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_rpkm</strong>: Pandas dataframe with R/FPKM-normalized</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
      <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="n">Gene1</span> <span class="mi">66</span>          <span class="mi">59</span>          <span class="mi">1</span>           <span class="mi">82</span>          <span class="mi">45</span>
<span class="n">Gene2</span> <span class="mi">35</span>          <span class="mi">0</span>           <span class="mi">7</span>           <span class="mi">72</span>          <span class="mi">2</span>
<span class="n">Gene3</span> <span class="mi">20</span>          <span class="mi">70</span>          <span class="mi">85</span>          <span class="mi">78</span>          <span class="mi">36</span>
<span class="n">Gene4</span> <span class="mi">96</span>          <span class="mi">7</span>           <span class="mi">93</span>          <span class="mi">38</span>          <span class="mi">85</span>
<span class="n">Gene5</span> <span class="mi">73</span>          <span class="mi">41</span>          <span class="mi">92</span>          <span class="mi">77</span>          <span class="mi">26</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">r_fpkm</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;/path/to/transcripts.gtf&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
      <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="n">Gene1</span> <span class="mf">15006.3436</span>  <span class="mf">21978.9881</span>  <span class="mf">237.1833</span>    <span class="mf">15581.6457</span>  <span class="mf">15294.6567</span>
<span class="n">Gene2</span> <span class="mf">18516.3632</span>  <span class="mf">0.0000</span>      <span class="mf">3863.1261</span>   <span class="mf">31833.8133</span>  <span class="mf">1581.6628</span>
<span class="n">Gene3</span> <span class="mf">1552.2985</span>   <span class="mf">8901.5987</span>   <span class="mf">6882.0428</span>   <span class="mf">5059.5089</span>   <span class="mf">4176.8031</span>
<span class="n">Gene4</span> <span class="mf">10220.8992</span>  <span class="mf">1221.0702</span>   <span class="mf">10328.8988</span>  <span class="mf">3381.1932</span>   <span class="mf">13527.9835</span>
<span class="n">Gene5</span> <span class="mf">12188.2602</span>  <span class="mf">11215.7274</span>  <span class="mf">16023.5923</span>  <span class="mf">10744.2995</span>  <span class="mf">6489.1599</span>
</pre></div>
</div>
</div>
<div class="section" id="tpm-transcripts-per-million">
<h2>TPM (Transcripts per Million)<a class="headerlink" href="#tpm-transcripts-per-million" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpressplot.tpm ( data, gtf, feature_type=’exon’, identifier=’gene_name’, sep=’t’ )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Perform transcripts per million sample normalization on RNAseq data</div>
<div class="line"><br /></div>
<div class="line">Formula:</div>
<div class="line"><span class="math notranslate nohighlight">\(TPM_{g} = \frac{1e6 \cdot r_{\textit{ge}}}{(\sum_{g=1}^{n} (\frac{1e3 \cdot r_{\textit{ge}}}{l_{\textit{ge}}})) \cdot \textit{l} _{\textit{ge}}}\)</span></div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Dataframe contains raw count data, where samples are along columns and genes across rows</div>
<div class="line">- By default, this will take the longest transcript based on combined exon length. If you prefer to use the cumulative exon length of the Ensembl canonical transcript, you must first curate the GTF using <code class="xref py py-data docutils literal notranslate"><span class="pre">xpresspipe</span> <span class="pre">modifyGTF</span> <span class="pre">-g</span> <span class="pre">transcripts.gtf</span> <span class="pre">-l</span></code></div>
<div class="line">- If you performed isoform quantification and wish to length normalize each isoform, provide <code class="xref py py-data docutils literal notranslate"><span class="pre">identifier='transcript_id'</span></code></div>
<div class="line">- If you would like to normalize based on CDS length, provide <code class="xref py py-data docutils literal notranslate"><span class="pre">feature_type='CDS'</span></code></div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe with counts values</div>
<div class="line"><strong>gtf</strong>: GTF reference file path and name</div>
<div class="line"><strong>feature_type</strong>: Label of feature to use in length normalization</div>
<div class="line"><strong>identifier</strong>: Label for how to group genes/transcripts. If normalizing for a gene, use <code class="xref py py-data docutils literal notranslate"><span class="pre">gene_name</span></code> or <code class="xref py py-data docutils literal notranslate"><span class="pre">gene_id</span></code>. If performing length normalization for isoforms, provide <code class="xref py py-data docutils literal notranslate"><span class="pre">transcript_id</span></code></div>
<div class="line"><strong>sep</strong>: GTF delimiter (usually tab-delimited)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_tpm</strong>: Pandas dataframe with TPM-normalized</div>
</div>
</div>
<div class="section" id="batch-normalize">
<h2>Batch Normalize<a class="headerlink" href="#batch-normalize" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpressplot.batch_normalize ( input_file, batch_file )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Control for batch effects between datasets</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for xpressplot usage where samples are normalized previously if desired</div>
<div class="line">- Requires a properly formatted dataframe complying to SVA COMBAT info file (see example below)</div>
<div class="line">- R is installed on your machine and is in your $PATH</div>
<div class="line">- All input files are tab-delimited (with .txt or .tsv suffix)</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>input_file</strong>: Input dataframe file with values (can be normalized or unnormalized)</div>
<div class="line"><strong>batch_file</strong>: Input dataframe containing batch effect information, column naming convention must be followed and <em>is</em> case-sensitive</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/expression.tsv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000227232</span> <span class="mf">66.34</span>   <span class="mf">59.13</span>   <span class="mf">1.90</span>    <span class="mf">82.49</span>
<span class="n">ENSG00000240361</span> <span class="mf">35.73</span>   <span class="mf">0.00</span>    <span class="mf">7.38</span>    <span class="mf">72.94</span>
<span class="n">ENSG00000238009</span> <span class="mf">20.02</span>   <span class="mf">70.21</span>   <span class="mf">85.10</span>   <span class="mf">78.87</span>
<span class="n">ENSG00000241860</span> <span class="mf">96.23</span>   <span class="mf">7.49</span>    <span class="mf">93.49</span>   <span class="mf">38.39</span>
<span class="n">ENSG00000187634</span> <span class="mf">73.91</span>   <span class="mf">41.28</span>   <span class="mf">92.27</span>   <span class="mf">77.93</span>
<span class="o">&gt;</span> <span class="n">batch</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/batch_info.tsv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">batch</span>
  <span class="n">Sample</span>  <span class="n">Batch</span>
<span class="mi">0</span> <span class="n">s1_rpf</span>  <span class="n">batch1</span>
<span class="mi">1</span> <span class="n">s1_rna</span>  <span class="n">batch2</span>
<span class="mi">2</span> <span class="n">s2_rpf</span>  <span class="n">batch1</span>
<span class="mi">3</span> <span class="n">s2_rna</span>  <span class="n">batch2</span>
<span class="o">&gt;</span> <span class="n">xp</span><span class="o">.</span><span class="n">batch_normalize</span><span class="p">(</span><span class="s1">&#39;/path/to/expression.tsv&#39;</span><span class="p">,</span> <span class="s1">&#39;/path/to/batch_info.tsv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="clean-data">
<h2>Clean Data<a class="headerlink" href="#clean-data" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpressplot.clean_df ( data, axis=0 )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Cleans NULL values from axis and clears duplicate indices</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for xpressplot usage</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe file with values (can be normalized or unnormalized)</div>
<div class="line"><strong>axis</strong>: Axis to clean NaN values from (default: 0, which corresponds to rows)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_clean</strong>: Cleaned pandas dataframe</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000227232</span> <span class="mf">66.34</span>   <span class="mf">59.13</span>   <span class="mf">1.90</span>    <span class="n">NA</span>
<span class="n">ENSG00000240361</span> <span class="mf">35.73</span>   <span class="mf">0.00</span>    <span class="mf">7.38</span>    <span class="mf">72.94</span>
<span class="n">Gene2</span>           <span class="mf">20.02</span>   <span class="mf">70.21</span>   <span class="mf">85.10</span>   <span class="mf">78.87</span>
<span class="n">Gene2</span>           <span class="mf">96.23</span>   <span class="mf">7.49</span>    <span class="mf">93.49</span>   <span class="mf">38.39</span>
<span class="n">ENSG00000187634</span> <span class="mf">73.91</span>   <span class="n">NA</span>      <span class="mf">92.27</span>   <span class="mf">77.93</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">clean_df</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000240361</span> <span class="mf">35.73</span>   <span class="mf">0.00</span>    <span class="mf">7.38</span>    <span class="mf">72.94</span>
</pre></div>
</div>
</div>
<div class="section" id="set-gene-threshold">
<h2>Set Gene Threshold<a class="headerlink" href="#set-gene-threshold" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpressplot.threshold ( data, minimum=None, maximum=None )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Cleans gene axis (assumed to by rows) of genes containing values below or above user-determined thresholds</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for xpressplot usage</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe file with values (can be normalized or unnormalized)</div>
<div class="line"><strong>minimum</strong>: Minimum value all samples need of a given gene to avoid dropping across all samples</div>
<div class="line"><strong>maximum</strong>: Maximum value all samples can have of a given gene to avoid dropping across all samples</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_clean</strong>: Cleaned pandas dataframe</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000227232</span> <span class="mf">66.34</span>   <span class="mf">59.13</span>   <span class="mf">1.90</span>    <span class="mf">82.49</span>
<span class="n">ENSG00000240361</span> <span class="mf">35.73</span>   <span class="mf">0.00</span>    <span class="mf">7.38</span>    <span class="mf">72.94</span>
<span class="n">ENSG00000238009</span> <span class="mf">20.02</span>   <span class="mf">70.21</span>   <span class="mf">85.10</span>   <span class="mf">78.87</span>
<span class="n">ENSG00000241860</span> <span class="mf">96.23</span>   <span class="mf">7.49</span>    <span class="mf">93.49</span>   <span class="mf">38.39</span>
<span class="n">ENSG00000187634</span> <span class="mf">73.91</span>   <span class="mf">41.28</span>   <span class="mf">92.27</span>   <span class="mf">77.93</span>
<span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">threshold</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">minimum</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_rpf</span>  <span class="n">s1_rna</span>  <span class="n">s2_rpf</span>  <span class="n">s2_rna</span>
<span class="n">ENSG00000238009</span> <span class="mf">20.02</span>   <span class="mf">70.21</span>   <span class="mf">85.10</span>   <span class="mf">78.87</span>
<span class="n">ENSG00000241860</span> <span class="mf">96.23</span>   <span class="mf">7.49</span>    <span class="mf">93.49</span>   <span class="mf">38.39</span>
<span class="n">ENSG00000187634</span> <span class="mf">73.91</span>   <span class="mf">41.28</span>   <span class="mf">92.27</span>   <span class="mf">77.93</span>
</pre></div>
</div>
</div>
<div class="section" id="prepare-xpressplot-dataset">
<h2>Prepare xpressplot Dataset<a class="headerlink" href="#prepare-xpressplot-dataset" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpressplot.prep_data ( data, info, gene_scale=True, print_means=False )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Prepare dataframe for downstream analyses</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for xpressplot usage (genes as rows, samples as columns)</div>
<div class="line">- Requires properly formatted xpressplot metadata dataframe</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: xpressplot formatted dataframe of expression values</div>
<div class="line"><strong>info</strong>: xpressplot formatted sample info dataframe</div>
<div class="line"><strong>gene_scale</strong>: Scale genes (rows) of data</div>
<div class="line"><strong>print_means</strong>: Print means for each sample verification</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_normalized</strong>: Normalized pandas dataframe</div>
<div class="line"><strong>data_labeled</strong>: Labeled pandas dataframe</div>
</div>
</div>
<div class="section" id="check-sample-expression-distributions">
<h2>Check Sample Expression Distributions<a class="headerlink" href="#check-sample-expression-distributions" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpressplot.check_samples ( data )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Visualize gene expression distributions on a sample-by-sample basis</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for xpressplot usage</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe file with values (can be normalized or unnormalized)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line">Boxplot with samples on the x-axis and lump expression distributions for all genes in that sample</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">xp</span><span class="o">.</span><span class="n">check_samples</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../_images/quality.png"><img alt="../_images/quality.png" class="align-left" src="../_images/quality.png" style="width: 576.0px; height: 288.0px;" /></a>
</div>
<div class="section" id="microarray-probe-collapse">
<h2>Microarray Probe Collapse<a class="headerlink" href="#microarray-probe-collapse" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>xpressplot.probe_collapse ( data, reference, gene_list=None, no_multimappers=True )</strong></div>
<div class="line"><br /></div>
<div class="line">Purpose:</div>
<div class="line">Remove multimapping probes and collapse probes mapping to the same gene by averaging the values for those probes per sample</div>
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- Requires a properly formatted dataframe for xpressplot usage</div>
<div class="line">- Assumes GPL .txt file from NCBI is tab delimited</div>
<div class="line"><br /></div>
</div>
<div class="line">Parameters:</div>
<div class="line"><strong>data</strong>: Input dataframe file with values (can be normalized or unnormalized)</div>
<div class="line"><br /></div>
<div class="line">Returns:</div>
<div class="line"><strong>data_collapsed</strong>: Pandas dataframe file probes collapsed and the corresponding gene names listed</div>
<div class="line"><br /></div>
<div class="line">Examples:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
          <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="mi">1007</span><span class="n">_s_at</span> <span class="mi">66</span>          <span class="mi">59</span>          <span class="mi">1</span>           <span class="mi">82</span>          <span class="mi">45</span>
<span class="mi">1053</span><span class="n">_at</span>   <span class="mi">35</span>          <span class="mi">0</span>           <span class="mi">7</span>           <span class="mi">72</span>          <span class="mi">2</span>
<span class="mi">121</span><span class="n">_at</span>    <span class="mi">20</span>          <span class="mi">70</span>          <span class="mi">85</span>          <span class="mi">78</span>          <span class="mi">36</span>
<span class="mi">218024</span><span class="n">_at</span> <span class="mi">96</span>          <span class="mi">7</span>           <span class="mi">93</span>          <span class="mi">38</span>          <span class="mi">85</span>
<span class="mi">240362</span><span class="n">_at</span> <span class="mi">73</span>          <span class="mi">41</span>          <span class="mi">92</span>          <span class="mi">77</span>          <span class="mi">26</span>
<span class="o">&gt;</span> <span class="n">probe_collapse</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">probe_collapse</span><span class="p">(</span><span class="n">probe_test</span><span class="p">,</span> <span class="s1">&#39;/path/to/gpl_ref.txt&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">probe_collapse</span>
      <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="n">MPC1</span>  <span class="mf">84.5</span>        <span class="mf">24.0</span>        <span class="mf">92.5</span>        <span class="mf">57.5</span>        <span class="mf">55.5</span>
<span class="n">PAX8</span>  <span class="mf">20.0</span>        <span class="mf">70.0</span>        <span class="mf">85.0</span>        <span class="mf">78.0</span>        <span class="mf">36.0</span>
<span class="n">RFC2</span>  <span class="mf">35.0</span>        <span class="mf">0.0</span>         <span class="mf">7.0</span>         <span class="mf">72.0</span>        <span class="mf">2.0</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span>
          <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="mi">1007</span><span class="n">_s_at</span> <span class="mi">66</span>          <span class="mi">59</span>          <span class="mi">1</span>           <span class="mi">82</span>          <span class="mi">45</span>
<span class="mi">1053</span><span class="n">_at</span>   <span class="mi">35</span>          <span class="mi">0</span>           <span class="mi">7</span>           <span class="mi">72</span>          <span class="mi">2</span>
<span class="mi">121</span><span class="n">_at</span>    <span class="mi">20</span>          <span class="mi">70</span>          <span class="mi">85</span>          <span class="mi">78</span>          <span class="mi">36</span>
<span class="mi">218024</span><span class="n">_at</span> <span class="mi">96</span>          <span class="mi">7</span>           <span class="mi">93</span>          <span class="mi">38</span>          <span class="mi">85</span>
<span class="mi">240362</span><span class="n">_at</span> <span class="mi">73</span>          <span class="mi">41</span>          <span class="mi">92</span>          <span class="mi">77</span>          <span class="mi">26</span>
<span class="o">&gt;</span> <span class="n">probe_collapse</span> <span class="o">=</span> <span class="n">xp</span><span class="o">.</span><span class="n">probe_collapse</span><span class="p">(</span><span class="n">probe_test</span><span class="p">,</span> <span class="s1">&#39;/path/to/gpl_ref.txt&#39;</span><span class="p">,</span> <span class="n">no_multimappers</span><span class="o">=</span><span class="bp">False</span><span class="p">))</span>
<span class="o">&gt;</span> <span class="n">probe_collapse</span>
                  <span class="n">fGSM523242</span>  <span class="n">fGSM523243</span>  <span class="n">fGSM523244</span>  <span class="n">fGSM523245</span>  <span class="n">fGSM523246</span>
<span class="n">DDR1</span> <span class="o">///</span> <span class="n">MIR4640</span>  <span class="mf">66.0</span>        <span class="mf">59.0</span>        <span class="mf">1.0</span>         <span class="mf">82.0</span>        <span class="mf">45.0</span>
<span class="n">MPC1</span>              <span class="mf">84.5</span>        <span class="mf">24.0</span>        <span class="mf">92.5</span>        <span class="mf">57.5</span>        <span class="mf">55.5</span>
<span class="n">PAX8</span>              <span class="mf">20.0</span>        <span class="mf">70.0</span>        <span class="mf">85.0</span>        <span class="mf">78.0</span>        <span class="mf">36.0</span>
<span class="n">RFC2</span>              <span class="mf">35.0</span>        <span class="mf">0.0</span>         <span class="mf">7.0</span>         <span class="mf">72.0</span>        <span class="mf">2.0</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="analysis.html" class="btn btn-neutral float-right" title="Analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="getting-data.html" class="btn btn-neutral float-left" title="Retrieving Data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>